### PR TITLE
Add support to export options in FBP files.

### PR DESCRIPTION
**Pull request by ibriano**
Make FBPs configurable by allowing them to export options from their
sub-nodes.

An options_description will be created for the type, copying the
option_description_member from the corresponding option of the node,
which means that required options from a sub-node will turn into
required options for the parent. Defaults are taken from the default
from the sub-node type, or the current value for that node's instance
if it has options.

### CODE REVIEW

**Comment by ceolin** [comment]
i've tested and this series looks good, although this is piece of code that i'm not familiar with.

+1

File: src/lib/flow/sol-flow-builder.c [hunk 1]
```diff
@@ -544,6 +616,15 @@ get_type_description(struct sol_flow_builder *builder)
         builder->type_desc.ports_out = builder->ports_out_desc.base.data;
     }
 
+    if (builder->options_description.len > 0) {
+        struct sol_flow_node_options_member_description *sentinel;
+        sentinel = sol_vector_append(&builder->options_description);
+        SOL_NULL_CHECK(sentinel, NULL);
+        memset(sentinel, 0, sizeof(*sentinel));
+        builder->type_desc.options = get_options_description(builder);
+        SOL_NULL_CHECK(builder->type_desc.options, NULL);
```
Comment (lpereira) [inline src/lib/flow/sol-flow-builder.c]: `sentinel` leaks here if the NULL-check fails. Is this OK?

Comment (ibriano) [inline src/lib/flow/sol-flow-builder.c]: Yes, if this fails, everything about this builder instance will fail and the whole vector will be freed.

File: src/lib/flow/sol-flow-builder.c [hunk 2]
```diff
@@ -581,6 +662,107 @@ get_exported_ports(
...
+        SOL_FLOW_NODE_OPTIONS_SUB_API_CHECK(copy_from, SOL_FLOW_BUILDER_API_VERSION, NULL);
+    }
+
+    opts = calloc(1, builder->options_size);
+    SOL_NULL_CHECK(opts, NULL);
+
+    opts->base.api_version = SOL_FLOW_NODE_OPTIONS_API_VERSION;
+    opts->base.sub_api = SOL_FLOW_BUILDER_API_VERSION;
+
+    for (member = type->description->options->members; member->name; member++) {
```
Comment (lpereira) [inline src/lib/flow/sol-flow-builder.c]: Holy indirection, Batman!

File: src/lib/flow/sol-flow-builder.c [hunk 3]
```diff
@@ -581,6 +662,107 @@ get_exported_ports(
...
+        dst = (char *)opts + member->offset;
+        if (copy_from)
+            src = (const char **)((char *)copy_from + member->offset);
+        else
+            src = (const char **)&member->defvalue.ptr;
+
+        if (is_string) {
+            char **s = (char **)dst;
+            free(*s);
+            *s = *src ? strdup(*src) : NULL;
```
Comment (lpereira) [inline src/lib/flow/sol-flow-builder.c]: If `is_string` and `strdup()` fails, will the called know how to handle `*s` being NULL?

File: src/lib/flow/sol-flow-builder.c [hunk 4]
```diff
@@ -955,3 +1143,97 @@ sol_flow_builder_export_out_port(struct sol_flow_builder *builder, const char *n
...
+        builder->options_size = sizeof(struct sol_flow_builder_options);
+
+    member_alignment = get_member_alignment(opt);
+    padding = builder->options_size % member_alignment;
+    exported_opt->offset = builder->options_size + padding;
+
+    builder->options_size += exported_opt->size + padding;
+    r = node_spec_add_options_reference(builder, node, exported_opt, opt);
+    if (r < 0) {
+        SOL_ERR("Failed to export option '%s' from node '%s'", option_name, node_name);
```
Comment (lpereira) [inline src/lib/flow/sol-flow-builder.c]: `exported_opt` leaks.

File: src/lib/flow/sol-flow-builder.c [hunk 5]
```diff
@@ -955,3 +1143,97 @@ sol_flow_builder_export_out_port(struct sol_flow_builder *builder, const char *n
 
     return 0;
 }
+
+static size_t
+get_member_alignment(const struct sol_flow_node_options_member_description *member)
+{
+    struct sol_str_slice t;
+    static const struct sol_str_table alignments[] = {
+        SOL_STR_TABLE_ITEM("boolean", __alignof__(member->defvalue.b)),
```
Comment (lpereira) [inline src/lib/flow/sol-flow-builder.c]: `__alignof__` is a GCC extension. Is it supported by all compilers we're also supporting?

Comment (ibriano) [inline src/lib/flow/sol-flow-builder.c]: It'd better be or nothing of this pull request is worth considering.

**Comment by cmarcelo** [comment]
Looking good. Two most important changes I would ask: to not include the builder in options and pipe it through the data type, and take out the foosball change -- explanation is there in the commit.

There's a test for declare that creates an adder that always adds 22. Could you add a variant of adder that takes the value to add from an option and feed it to a constant inside? This would complement the test you already added (that just checks for the members being there and from the same type).

Great work!

**Comment by ibriano** [comment]
New version in #150

**Pull request closed without merging**